### PR TITLE
Create Posts Method

### PR DESCRIPTION
**Pull request by MattsAttack**
## Description

This PR allows the user to create new posts with the fab on the feeds page.

_Closes: #5_

## Type of Change

- ✨ New feature

## Checklist

> Before submitting the PR, please make sure you do the following:

- [ ] Read the Contributing Guidelines.
- [ ] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [ ] Fill out this template.
- [ ] Log your hours.
- [ ] Check that commits follow the Angular commit convention, more or less.
- [ ] Ideally, include relevant tests that fail without this PR but pass with it (if applicable).

## Tested on

> What device(s) have you tested this on?
> Delete all non-applicable examples.

- CI
- macOS 14
- Ubuntu 24.04
- Windows 10
- etc.

### CODE REVIEW

File: packages/app/test/src/features/home/application/feed_service_test.dart [hunk 1]
```diff
@@ -25,10 +26,14 @@ void main() {
 
       // Act
       final postsToAdd = List.generate(
-        pageSize,
-        (index) => const PostEntity(
+        pageSize, //@lishaduck implemented a fix here but you may want to review
```
Comment (lishaduck) [inline packages/app/test/src/features/home/application/feed_service_test.dart:29]: > What fix?
> 
> 
> 
> Also, please make a draft PR. 

 _Originally posted by @lishaduck in [c090f45](https://github.com/PHS-TSA/nexus/commit/c090f457a414d4df05404c2071fb26e684dcc5f2#r147682001)_

File: packages/app/lib/src/features/home/domain/post_entity.dart [hunk 2]
```diff
@@ -16,10 +19,24 @@ sealed class PostEntity with _$PostEntity {
...
   const factory PostEntity({
+    /// The title of the post.
+    required String headline,
+
     /// The textual content of the post.
-    required String body,
+    required String description,
+
+    /// The author of the post.
+    required UserId author,
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/domain/post_entity.dart:29]: > @lishaduck do we want classify users by their user id or email for posts? As of now I have it by email in appwrite but the posts class does it by user id. Which way do you think is gonna be better? 

 _Originally posted by @MattsAttack in [00dd7c4](https://github.com/PHS-TSA/nexus/commit/00dd7c4ae5426e55f6302e4a5bfe185e264ecf92#commitcomment-147671448)_

> I think user ID is probably a more robust solution. 

 _Originally posted by @lishaduck in [00dd7c4](https://github.com/PHS-TSA/nexus/commit/00dd7c4ae5426e55f6302e4a5bfe185e264ecf92#commitcomment-147678094)_

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+/*
+How to do things:
+Put database and collection id in .env
+Implement post code from other branch
+For first implementation have post_repository read in all posts and then add queries and other filters down the line
+Have flutter access user location
+*/
+
+// Might need to have 2 posts lists. One for local and one for world
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:21]: Should just need one list (in the service), but multiple providers.

See: [Provider](https://riverpod.dev/docs/providers/provider) and [Combining Provider States](https://riverpod.dev/docs/concepts/combining_providers). There used to be better docs, but I can't find them.

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  List<Document> _posts = [];
+
+  //get all todos func for other classes
+  List<Document> get allPosts => _posts;
+
+  bool _isLoading = false;
+
+  //check if loading
+
+  bool get checkLoading => _isLoading;
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:44]: State belongs in a provider. Repos should be purely functional and stateless.

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+  Future createNewPost(
+    String? headline,
+    String? description,
+    UserId author,
+    LatLng location,
+    ImageProvider? image,
+  ) async {
+    // Maybe change to no author param and getting value in here. Whatever's more efficient
+    //TODO get current users email or user ID (talk to eli about which one)
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:78]: If the author is always current user, then yes. We probably want to cache it though, so no?
I recall changing my mind in [`PSDTools/app`](https://github.com/PSDTools/app), so check the [`CoinsRepository`](https://github.com/PSDTools/app/blob/main/lib/features/pirate_coins/data/coins_repository.dart) over there.

**Review by lishaduck — COMMENTED** [review]
Responded to your code comments and moved stuff over from the commit comment pages.

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 6]
```diff
@@ -24,6 +24,13 @@ Have flutter access user location
 
 abstract interface class PostRepository {
   //Have funcs in heres
+
+  // TODOadd in get funcs
+
+  final List<Document> _posts = [];
+
+  //get all todos func for other classes
+  List<Document> get allPosts => _posts;
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:33]: You really don't need this. Repos have funcs. Services have state like this. We can talk about it on Thurs.

File: packages/app/lib/src/features/home/data/post_repository.dart (hunk 5 above)
Comment (lishaduck) [reply packages/app/lib/src/features/home/data/post_repository.dart:89]: Answer: I ended up passing it in as an argument to the repo. Note that to do that, you'd want to get rid of state in the repo (which is necessary either way) so that you don't run the risk of a singleton.

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 7]
```diff
@@ -71,6 +72,16 @@ final class _AppwritePostRepository implements PostRepository {
...
 
+    print('returning docs');
+    print(documentList.documents[1].data);
+    // print(
+    //   documentList.documents
+    //       .map((document) => PostEntity.fromJson(document.data))
+    //       .toList(),
+    // );
+
+    // @lishaduck after some debugging, I'm pretty sure there's an issue with the json mapping to PostEntitys. Its getting the documents but it seems like its just not mapping the values correctly
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:83]: Hmmm. Is it throwing, or returning an empty list, or what?

File: packages/app/lib/src/app/wrapper_page.dart [hunk 8]
```diff
@@ -99,4 +103,94 @@ class WrapperPage extends ConsumerWidget {
       },
     );
   }
+
+  Future<void> _dialogBuilder(BuildContext context) {
```
Comment (lishaduck) [inline packages/app/lib/src/app/wrapper_page.dart:107]: Extract it out to a private widget, i.e.:

```dart
class _DialogBuilder extends HookWidget {/*...*/}
```

File: packages/app/lib/src/app/wrapper_page.dart [hunk 9]
```diff
@@ -99,4 +103,94 @@ class WrapperPage extends ConsumerWidget {
...
+    // final formKey = useMemoized(GlobalKey<FormState>.new);
+
+    // final headline = useState('');
+    // final description = useState('');
+    // TODOadd images
+
+    TextEditingController titleController = TextEditingController();
+    TextEditingController descriptionController = TextEditingController();
```
Comment (lishaduck) [inline packages/app/lib/src/app/wrapper_page.dart:115]:
```suggestion
    final titleController = useTextEditingController();
    final descriptionController = useTextEditingController();
```

**Review by lishaduck — CHANGES_REQUESTED** [review]

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 10]
```diff
@@ -31,7 +31,7 @@ abstract interface class PostRepository {
   /// Create a new post.
   ///
   /// Returns the created post.
-  Future<PostEntity> createNewPost(
+  Future<void> createNewPost(
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:35]: Why this? It's useful to be able to get the post back out.

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 11]
```diff
@@ -46,6 +46,7 @@ final class _AppwritePostRepository implements PostRepository {
     required this.databaseId,
     required this.collectionId,
     required this.author,
+    // required this.feed,
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:50]: Why did you remove this, was it just becuase it was unused? (fair enough, just curious)

File: packages/app/lib/src/app/wrapper_page.dart (hunk 8 above)
Comment (MattsAttack) [reply packages/app/lib/src/app/wrapper_page.dart]: How would i implement this. I'm a little confused on what you mean here

File: packages/app/lib/src/features/home/data/post_repository.dart (hunk 11 above)
Comment (lishaduck) [reply packages/app/lib/src/features/home/data/post_repository.dart:50]: You'd wanted to do conditional queries with it, but I'm always happy to delete dead code. I just thought you'd asked for it.

File: packages/app/lib/src/features/home/data/post_repository.dart [hunk 12]
```diff
@@ -108,7 +109,7 @@ final class _AppwritePostRepository implements PostRepository {
           'author': author,
           'lat': lat,
           'lng': lng,
-          'timestamp': DateTime.timestamp(),
+          'timestamp': DateFormat('yyyy-MM-ddTHH:mm:ss').format(DateTime.now()),
```
Comment (lishaduck) [inline packages/app/lib/src/features/home/data/post_repository.dart:112]: I still feel an extension type would be more robust.

Comment (lishaduck) [reply packages/app/lib/src/features/home/data/post_repository.dart:112]: On second thought, I forgot about [`JsonConverter`s](https://github.com/google/json_serializable.dart/blob/3e4dbe3ff6b7989e99e835a6533ce6fcfb2a8e51/example/lib/json_converter_example.dart#L26-L34), we should swap out the unneeded extension types with them.

Comment (lishaduck) [reply packages/app/lib/src/features/home/data/post_repository.dart:112]: Also, DateTime.now isn't in UTC.

File: packages/app/lib/src/features/home/data/post_repository.dart (hunk 11 above)
Comment (MattsAttack) [reply packages/app/lib/src/features/home/data/post_repository.dart:50]: Sorry, not that I can remember. I can't think of any feature that would need the data of the post the user just created.

Comment (lishaduck) [reply packages/app/lib/src/features/home/data/post_repository.dart:50]: This comment was about passing the feed to the repo, not returning a post from createNewPost.

File: pubspec.yaml [hunk 13]
```diff
@@ -6,3 +6,5 @@ environment:
 
 dev_dependencies:
   melos: ^6.1.0
+dependencies:
+  geolocator: ^13.0.1
```
Comment (lishaduck) [inline pubspec.yaml:10]: First, why?
Second, wrong pubspec.

Comment (lishaduck) [reply pubspec.yaml]: The # of conflicts I'm gonna have when I push...

P.S. (You removed it, but didn't add it to the other one)

Comment (MattsAttack) [reply pubspec.yaml]: It's in the pubspec.yaml in the app folder with the other dependencies.
image

Comment (lishaduck) [reply pubspec.yaml]: Huh. Ok then.
